src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -946,7 +946,8 @@ ss::future<
   result<async_manifest_view::archive_start_offset_advance, error_outcome>>
 async_manifest_view::compute_retention(
   std::optional<size_t> size_limit,
-  std::optional<std::chrono::milliseconds> time_limit) noexcept {
+  std::optional<std::chrono::milliseconds> time_limit,
+  std::optional<kafka::offset> pinned_offset) noexcept {
```

Comment:
This change makes retention calculation non-idempotent because it depends on the parameter now. 

Few questions:
- if the data is about to be deleted by the retention, is it really necessary to translate it and send to iceberg?
- wouldn't it be sufficient to just pause cloud retention while translation is trying to keep up?

Reply:
>if the data is about to be deleted by the retention, is it really necessary to translate it and send to iceberg?

The issue is that users don't really have much control over how fast or slow translation is, or when it gets scheduled. Given that, it's hard to claim any kind of delivery guarantees if we allow data to be removed before translating

>wouldn't it be sufficient to just pause cloud retention while translation is trying to keep up?

Yea that's true. Not a bad idea if trying to get this in quick. I may consider doing this, just to avoid polluting the retention code for now. Though it's certainly a hack -- long term, I think we'll need to update retention to acknowledge the pinned offset.

Reply:
I went the route of just adding unit tests for the relevant subsystems. I'm open to more discussion towards your approach if you think this is still too complex.

I also removed the change from the compaction code and instead relied on the existing compaction config parameter.